### PR TITLE
✨ [source-zendesk] upgrade to CDK v6 to use concurrency for incremental streams

### PR DESCRIPTION
**Pull request by brianjlai**
NOTE: This PR also includes the CDK changes needed to fix a few bugs related to stream state converting. I need them in the branch to build a version of zendesk with the local cdk. Once the changes are vetted with regression testing, I will create a separate PR to release the CDK changes

## What

Upgrades `source-zendesk-support` to the latest CDK 

## How

Updated run.py, source.py to take in the required arguments

replace `stream_state` with `stream_interval` in manifest which is thread safe

add concurrency level based on the API docs from zendesk

fix tests (i still need to fix 3 more)

todo: look into regression test failures

## User Impact

Potential rate limiting since we are syncing at a faster rate by default

## Can this PR be safely reverted and rolled back?

We still emit state in the legacy format to allow for rollbacks

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/datetime_stream_state_converter.py [hunk 1]
```diff
@@ -183,7 +183,7 @@ def __init__(
         self._parser = DatetimeParser()
 
     def output_format(self, timestamp: datetime) -> str:
-        return timestamp.strftime(self._datetime_format)
+        return self._parser.format(timestamp, self._datetime_format)
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converters/datetime_stream_state_converter.py:186]: Similar to https://github.com/airbytehq/airbyte/pull/48361, although it adds a little bit of a dependency on the declarative framework, it's safer to retain the exact behavior by using the parser. And we can always move this code into concurrent cursor later

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py [hunk 2]
```diff
@@ -1,14 +1,48 @@
...
+from typing import List
 
-from airbyte_cdk.entrypoint import launch
+from airbyte_cdk.entrypoint import AirbyteEntrypoint, launch, logger
+from airbyte_cdk.exception_handler import init_uncaught_exception_handler
+from airbyte_cdk.models import AirbyteErrorTraceMessage, AirbyteMessage, AirbyteTraceMessage, TraceType, Type
 from source_zendesk_support import SourceZendeskSupport
 
 
+def _get_source(args: List[str]):
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py:16]: It feels like this method should be imported and reused from CDK, rather than be applied to each source / destination separately.

Do we consider this in the future?

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py [hunk 3]
```diff
@@ -1,14 +1,48 @@
...
+        )
+        return None
+
+
 def run():
-    source = SourceZendeskSupport()
-    launch(source, sys.argv[1:])
+    init_uncaught_exception_handler(logger)
+    _args = sys.argv[1:]
+    source = _get_source(_args)
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py:46]: `_init_source` is a better name for this, wdyt?

**Comment by maxi297** [comment]
Here are the CDK changes to support the change: https://github.com/airbytehq/airbyte/pull/48389

Here is a printscreen of the locally run regression tests:
image

File: airbyte-integrations/connectors/source-zendesk-support/acceptance-test-config.yml [hunk 4]
```diff
@@ -39,7 +39,7 @@ acceptance_tests:
       - config_path: "secrets/config.json"
         configured_catalog_path: "integration_tests/incremental_catalog.json"
         future_state:
-          future_state_path: "integration_tests/abnormal_state.json"
+          bypass_reason: "This test does not make sense using Concurrent CDK"
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-support/acceptance-test-config.yml:42]: This however no longer holds true in the concurrent framework because the final state we emit is either latest synced record of a date window, or the lower bound of the window. We don’t want to use the upper bound of the window because its possible we’ve missed delayed records in that window. This was fixed in the [PR](https://github.com/airbytehq/airbyte/pull/45180/files#diff-e1414d3b2f03677efb789279bea1737ab2437e04bd3c3afe3d817656220ccf86R56-R57). And this will now lead to the mentioned test always failing because instead of emitting the date impossibly far in the future, we emit the bottom of the window which is typically the start date.

We intend to delete this test eventually anyway. For more information, see https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1730329381148769

File: airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py [hunk 5]
```diff
@@ -159,26 +152,25 @@ def test_default_start_date():
...
     # we expect base64 from creds input
-    result = SourceZendeskSupport().get_authenticator(config=config)
+    result = SourceZendeskSupport(config=config, catalog=None, state=None).get_authenticator(config=config)
     assert result._token == expected
 
 
 @pytest.mark.parametrize(
     "response, start_date, check_passed",
-    [({"active_features": {"organization_access_enabled": True}}, "2020-01-01T00:00:00Z", True), ({}, "2020-01-00T00:00:00Z", False)],
+    [([{"active_features": {"organization_access_enabled": True}}], "2020-01-01T00:00:00Z", True), ([], "2020-01-01T00:00:00Z", False)],
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py:166]: Changed two things:
* Mocked `UserSettingsStream.read_records` instaed of `UserSettingsStream.get_settings` to have better test coverage
* Fixed the datetime `2020-01-00T00:00:00Z` that is not a valid date

File: airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py [hunk 6]
```diff
@@ -159,26 +152,25 @@ def test_default_start_date():
     [
         (TEST_CONFIG, "aW50ZWdyYXRpb24tdGVzdEBhaXJieXRlLmlvL3Rva2VuOmFwaV90b2tlbg=="),
         (TEST_CONFIG_OAUTH, "test_access_token"),
-        (TEST_OLD_CONFIG, "aW50ZWdyYXRpb24tdGVzdEBhaXJieXRlLmlvL3Rva2VuOmFwaV90b2tlbg=="),
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py:162]: This one is interesting. Using the latest version of the source and this config:
```
{
    "subdomain": "d3v-airbyte",
    "start_date": "2022-06-01T00:00:00Z",
    "ignore_pagination": true,
    "auth_method": {
        "auth_method": "api_token",
        "email": <redacted>,
        "api_token": <redacted>
    }
}
```

... I would get the following error during a read:
```
{
    "type": "TRACE",
    "trace": {
        "type": "ERROR",
        "emitted_at": 1731073703195,
        "error": {
            "message": "Something went wrong in the connector. See the logs for more details.",
            "internal_message": "The path from `authenticator_selection_path` is not found in the config.",
            "stack_trace": "Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/auth/selective_authenticator.py\", line 30, in __new__
    selected_key = str(dpath.get(config, authenticator_selection_path))
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/dpath/__init__.py\", line 192, in get
    raise KeyError(glob)
KeyError: ['credentials', 'credentials']

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/main.py\", line 8, in <module>
    run()
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py\", line 14, in run
    launch(source, sys.argv[1:])
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 263, in launch
    for message in source_entrypoint.run(parsed_args):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 133, in run
    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/entrypoint.py\", line 191, in read
    for message in self.source.read(self.logger, config, catalog, state):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 165, in read
    yield from super().read(logger, config, catalog, state)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/abstract_source.py\", line 98, in read
    stream_instances = {s.name: s for s in self.streams(config)}
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py\", line 176, in streams
    declarative_streams = super().streams(args)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 95, in streams
    source_streams = [
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 96, in <listcomp>
    self._constructor.create_component(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 292, in create_component
    return self._create_component_from_model(model=declarative_component_model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 738, in create_declarative_stream
    retriever = self._create_component_from_model(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 1224, in create_simple_retriever
    requester = self._create_component_from_model(model=model.requester, decoder=decoder, config=config, name=name)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 897, in create_http_requester
    self._create_component_from_model(model=model.authenticator, config=config, url_base=model.url_base, name=name, decoder=decoder)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 300, in _create_component_from_model
    return component_constructor(model=model, config=config, **kwargs)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 1186, in create_selective_authenticator
    return SelectiveAuthenticator(  # type: ignore[abstract]
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/auth/selective_authenticator.py\", line 32, in __new__
    raise ValueError(\"The path from `authenticator_selection_path` is not found in the config.\") from err
ValueError: The path from `authenticator_selection_path` is not found in the config.
",
            "failure_type": "system_error"
        }
    }
}
```

The config has been updated in our GSM on `3/15/22, 7:49 AM`. I would assume it was because of [this change](https://github.com/airbytehq/airbyte/pull/11162/files#diff-2f6a32b67d7c88a2a4ccbb67200caeb69066953dea81994ec68a171414a11d0d). However, the manifest.yaml never took into account the old config format [see first commit [here](https://github.com/airbytehq/airbyte/pull/36823/files#diff-40506820d30657ee0d53226799b3489cf4cc350afe6147ee46c3efb551237bf7R15-R21)). Hence, I assume that old configs don't work since the introduction of manifest.yaml for any source that was low-code.

Why is this breaking now? It is because we call `streams` in `__init__` now which instantiate those streams in order to dispatch them as `concurrent_streams` or `sequential_streams`. We've made this conscious decision with @brianjlai knowing the consequences because else, the variable `concurrent_streams` and `sequential_streams` would have been stateful (sometimes instantiated, some other times not) which would have had complexity in `ConcurrentDeclarativeSource`

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py [hunk 7]
```diff
@@ -43,8 +43,12 @@ def __init__(self, email: str, password: str):
 
 
 class SourceZendeskSupport(YamlDeclarativeSource):
-    def __init__(self):
-        super().__init__(**{"path_to_yaml": "manifest.yaml"})
+    def __init__(self, catalog: Optional[ConfiguredAirbyteCatalog], config: Optional[Mapping[str, Any]], state: TState, **kwargs):
+        # Before 2024-11-11, the config was being modified in `streams`. We can't do that anymore because `ConcurrentDeclarativeSource` use
+        # the config to make concurrent components. Hence, the "main" config needs to be the declarative one and if Python sources need
+        # something a bit different, it needs to deduce it from the declarative config
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py:49]: Without those lines, we would get the following error:
`` 
{
    "type": "TRACE",
    "trace": {
        "type": "ERROR",
        "emitted_at": 1731074703862,
        "error": {
            "message": "Error starting the sync. This could be due to an invalid configuration or catalog. Please contact Support for assistance. Error: time data 'None' does not match format '%Y-%m-%dT%H:%M:%SZ'",
            "stack_trace": "Traceback (most recent call last):
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py\", line 22, in _get_source
    return SourceZendeskSupport(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py\", line 47, in __init__
    super().__init__(catalog=catalog, config=config, state=state, **{\"path_to_yaml\": \"manifest.yaml\"})
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/yaml_declarative_source.py\", line 31, in __init__
    super().__init__(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 65, in __init__
    self._concurrent_streams, self._synchronous_streams = self._group_streams(config=config or {})
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/concurrent_declarative_source.py\", line 164, in _group_streams
    cursor, connector_state_converter = self._constructor.create_concurrent_cursor_from_datetime_based_cursor(
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py\", line 564, in create_concurrent_cursor_from_datetime_based_cursor
    start_date = interpolated_start_date.get_datetime(config=config)
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/min_max_datetime.py\", line 58, in get_datetime
    time = self._parser.parse(str(self.datetime.eval(config, **additional_parameters)), datetime_format)  # type: ignore # datetime is always cast to an interpolated string
  File \"/Users/maxime/devel/code/airbyte/airbyte-integrations/connectors/source-zendesk-support/.venv/lib/python3.10/site-packages/airbyte_cdk/sources/declarative/datetime/datetime_parser.py\", line 35, in parse
    parsed_datetime = datetime.datetime.strptime(str(date), format)
  File \"/Users/maxime/.pyenv/versions/3.10.11/lib/python3.10/_strptime.py\", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File \"/Users/maxime/.pyenv/versions/3.10.11/lib/python3.10/_strptime.py\", line 349, in _strptime
    raise ValueError(\"time data %r does not match format %r\" %
ValueError: time data 'None' does not match format '%Y-%m-%dT%H:%M:%SZ'
"
        }
    }
}
```

The reason is that the `start_date` is optional in the config but the child stream (`SourceZendeskSupport` in this case) modify the config to instantiate the declarative streams, `ConcurrentDeclarativeSource` do not see those changes and this might cause errors when grouping the streams as part of `ConcurrentDeclarativeSource.__init__`.

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py [hunk 8]
```diff
@@ -115,13 +114,12 @@ def convert_config2stream_args(cls, config: Mapping[str, Any]) -> Mapping[str, A
...
     def convert_config_to_declarative_stream_args(cls, config: Mapping[str, Any]) -> Mapping[str, Any]:
-        """Convert input configs to parameters of the future streams
-        This function is used by unit tests too
+        """
+        Sets defaults to `start_date` and `ignore_pagination`.
         """
         return {
             "subdomain": config["subdomain"],
             "start_date": config.get("start_date", cls.get_default_start_date()),
-            "auth_type": config.get("auth_type"),
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py:124]: I didn't see any reference to `auth_type` in the manifest hence why I removed this

File: airbyte-integrations/connectors/source-zendesk-sunshine/integration_tests/abnormal_state.json [hunk 9]
```diff
@@ -1,5 +0,0 @@
-{
```
Comment (brianjlai) [inline airbyte-integrations/connectors/source-zendesk-sunshine/integration_tests/abnormal_state.json:1]: @maxi297 I think you mean to delete the `abnormal_state.json` file for `source-zendesk-support` instead of sunshine haha

Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-sunshine/integration_tests/abnormal_state.json:1]: Oops! I'm clearly not the sharpest shooter in the west

File: airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py (hunk 6 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py:162]: Thanks for the detailed writeup, i think we can just accept that if a connector is still using a 2 1/2 year old config, we don't have to guarantee it's going to work and we can accept this type of failure might exist

File: airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py (hunk 5 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/unit_tests/unit_test.py:166]: `2020-01-00T00:00:00Z` == `2019-12-31T00:00:00Z`. everybody knows this

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py (hunk 7 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py:49]: What still isn't obvious to me is why we have to inject the optional 2 year start date in code.

I'm assuming that this is failing for the semi incremental cursor here:
```
incremental_sync:
  type: DatetimeBasedCursor
  cursor_datetime_formats:
    - "%Y-%m-%dT%H:%M:%SZ"
    - "%Y-%m-%dT%H:%M:%S%z"
  datetime_format: "%Y-%m-%dT%H:%M:%SZ"
  cursor_field: "{{ parameters.get('cursor_field',  'updated_at') }}"
  start_datetime:
    datetime: "{{ config.get('start_date') }}"
```

I feel like we should be able to just add a condition to the interpolation to look back 2 years which simplify the python code that's required. Something like this. I'll experiment with why we didn't just do this instead of this extra config transformation logic

```
start_datetime:
  datetime: "{{ config.get('start_date') or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"

```

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py (hunk 2 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/run.py:17]: yes i think it would make sense given the repetition thusfar

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py (hunk 7 above)
Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/source.py]: noting that i removed this block of code because we can actually account for this using interpolation instead of custom python code:

in `semi_incremental_stream.incremental_stream`:
`"{{ config.get('start_date') or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"`

If `start_date` is missing, then we'll default to two years previously, and i tested this removing it from a config and the sync will not experience this error

**Comment by brianjlai** [comment]
Regression test results analysis:

After a few bug fixes and analyzing the latest results, there are mismatches. However, after inspecting the results and doing some local testing, these look to be expected mismatches.

The main issue w/ the current regression results is because new data for the target is added so rapidly, we rarely get the exact count to align since streams are no longer run at almost the same time. I’ve added some notes during my testing. The main manual validation I did was adding in a end date filter to stop new records from causing mismatched counts

Of the streams with mismatches that have been converted to concurrent streams:

- ticket_comments - I spot checked this stream using a start date `2024-11-20T02:10:03Z` and end date filter `2024-11-20T02:30:00Z`, this resulted in the same number of records on both control and target. It also used a custom record_selector
- ticket_metric_events_stream - very high volume, we expect a lot of changes or record falling out of the range
- ticket_audits - i suspect the same here, but this is a semi-incrmeental data_feed so i will do one last validation
- users - Same manual testing as `ticket_comments` and the counts lined up. I also verified that the first 5 and last 5 records are the same so we are requesting the correct range

Stream mismatches for other streams:

- tickets - This is a python defined stream. I suspect that mismatches are a result of this stream running at a different time than before. Since the concurrent framework syncs all concurrent streams before starting synchronous ones, the ordering will now different and it’s possible `tickets` is run earlier on the control version. I spot checked this and as expected, the high volume between runs leads to more record if run later. 
- ticket_metrics - same as above
- tags - full refresh so not run concurrently

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml [hunk 10]
```diff
@@ -87,7 +87,7 @@ definitions:
       datetime_format: "%Y-%m-%dT%H:%M:%SZ"
       cursor_field: "{{ parameters.get('cursor_field',  'updated_at') }}"
       start_datetime:
-        datetime: "{{ config.get('start_date') }}"
+        datetime: "{{ config.get('start_date') or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml:90]: TIL about using `day_delta` instead of hardcoding a default startdate. /cc @btkcodedev this is a cool trick!

File: airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml [hunk 11]
```diff
@@ -98,7 +98,7 @@ definitions:
     datetime_format: "%s"
     cursor_field: "{{ parameters.get('cursor_field',  'updated_at') }}"
     start_datetime:
-      datetime: "{{ timestamp(config.get('start_date')) | int or day_delta(-730, '%Y-%m-%dT%H:%M:%SZ') }}"
+      datetime: "{{ timestamp(config.get('start_date')) | int if config.get('start_date') else day_delta(-730, '%s') }}"
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml:101]: Weird, did this bug out previously? Seems unrelated to concurrency work.

Comment (brianjlai) [reply airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/manifest.yaml:101]: we need to do this because i got rid of the extra python code `convert_config_to_declarative_stream_args()` so we're more declcarative and have less custom code.

And because we no longer automatically inject `start_date`, we can't just use `or` because the `None | int` would fail types or something along those lines.

**Review by maxi297 — APPROVED** [review]
LGTM! Thanks for being so diligent on this @brianjlai !
